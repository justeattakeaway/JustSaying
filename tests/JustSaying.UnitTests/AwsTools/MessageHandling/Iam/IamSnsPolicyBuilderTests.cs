using JustSaying.AwsTools.MessageHandling;
using Newtonsoft.Json.Linq;

namespace JustSaying.UnitTests.AwsTools.MessageHandling.Iam;

public class IamSnsPolicyBuilderTests
{
    [Fact]
    public void ShouldGenerateApprovedIamPolicy()
    {
        // arrange
        var snsPolicyDetails = new SnsPolicyDetails
        {
            SourceArn = "arn:aws:sns:ap-southeast-2:123456789012:topic",
            AccountIds = new[] { "123456789012" }
        };

        // act
        var policy = IamSnsPolicyBuilder.BuildPolicyJson(snsPolicyDetails);

        // assert
        policy.ShouldMatchApproved(c =>
        {
            c.SubFolder("Approvals");
            c.WithScrubber(ScrubSids);
        });
    }

    [Theory]
    [InlineData("")]
    [InlineData(null)]
    [InlineData("arn:aws:service:region:123456789012")]// missing topic
    public void ShouldThrowArgumentExceptionWhenUsingInvalidArn(string sourceArn)
    {
        // arrange
        var snsPolicyDetails = new SnsPolicyDetails
        {
            SourceArn = sourceArn
        };

        // Act + Assert
        Should.Throw<ArgumentException>(() => IamSnsPolicyBuilder.BuildPolicyJson(snsPolicyDetails));
    }

    private static string ScrubSids(string iamPolicy)
    {
        // Sids are generated from guids on each invocation so must be ignored
        // when performing approval tests
        var json = JObject.Parse(iamPolicy);
        return iamPolicy
            .Replace(json["Statement"]![0]!["Sid"]!.ToString(), "<sid1>")
            .Replace(json["Statement"]![1]!["Sid"]!.ToString(), "<sid2>");
    }
}
