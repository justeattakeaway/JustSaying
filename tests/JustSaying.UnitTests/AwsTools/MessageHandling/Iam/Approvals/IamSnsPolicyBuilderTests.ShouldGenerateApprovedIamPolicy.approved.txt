{
    "Version" : "2012-10-17",
    "Statement" : [
        {
            "Sid" : "<sid1>",
            "Effect" : "Allow",
            "Principal" : {
                "AWS" : "*"
            },
            "Action"    : [
                "sns:GetTopicAttributes",
                "sns:SetTopicAttributes",
                "sns:AddPermission",
                "sns:RemovePermission",
                "sns:DeleteTopic",
                "sns:Subscribe",
                "sns:Publish"
            ],
            "Resource"  : "arn:aws:sns:ap-southeast-2:123456789012:topic",
            "Condition" : {
                "StringEquals" : {
                    "AWS:SourceOwner" : "123456789012"
                }
            }
        },
        {
            "Sid" : "<sid2>",
            "Effect" : "Allow",
            "Principal" : {
                "AWS" : ["123456789012"]
            },
            "Action"    : "sns:Subscribe",
            "Resource"  : "arn:aws:sns:ap-southeast-2:123456789012:topic"
        }
    ]
}